### DISCUSSION TITLE
How to output variables in source codes to exodus file

### DISCUSSION THREAD

**Discussion by pshen20127** [idaholab/moose#21585] [Q&A Modules: Solid mechanics] [answered]
Hello all

I hope to output a variable 'equivalent_slip_increment' in the file 'ComputeMultipleCrystalPlasticityStress.C' when running the simulation. I directly add a AuxVariable in input file as below. But the output value is zero. What should I do to fix it? Thanks. (Input file 'polycrystal.txt' is attached)
![微信截图_20220713214743](https://user-images.githubusercontent.com/81499506/178896033-b8403a19-b41c-4a96-9cec-b7c1f65470e2.png)
![微信截图_20220713214813](https://user-images.githubusercontent.com/81499506/178896047-dafb5bbe-f4a1-4446-9443-bf013a3437fe.png)
![微信截图_20220713214839](https://user-images.githubusercontent.com/81499506/178896055-013e6ae5-9f2a-47de-a972-1c4e1f318ddc.png)
![微信截图_20220713215212](https://user-images.githubusercontent.com/81499506/178896069-e969b5d0-2cb8-4a82-babf-994476a21cfb.png)
[polycrystal.txt](https://github.com/idaholab/moose/files/9108330/polycrystal.txt)

**Comment by sapitts** [comment, marked as answer]
Hi @pshen20127,

The Rank-2 tensor `equivalent_slip_increment` is a local variable and the `RankTwoAux` auxvariable works with those Rank-2 tensors that are declared as material properties. See the declaration of the `plastic_deformation_gradient` in `ComputeMultiCrystalPlasticityStress` here: https://github.com/idaholab/moose/blob/3bb5876c5c3f29649f426ff5535f8a3007658acb/modules/tensor_mechanics/src/materials/crystal_plasticity/ComputeMultipleCrystalPlasticityStress.C#L82 

If you would like to monitor the change in plastic strain, the `plastic_deformation_gradient` may be a good option. If you would like to monitor the slip increment on individual slip systems, the vector `slip_increment`would be a good option. Use the AuxKernel `MaterialStdVectorAux` as shown in this regression test input file: https://github.com/idaholab/moose/blob/next/modules/tensor_mechanics/test/tests/crystal_plasticity/stress_update_material_based/update_method_test.i

Finally, please post input file snippets as quoted text, not as screenshots, as outlined in the posting guidelines here: https://github.com/idaholab/moose/discussions/18270

Hope this helps!

**Reply by pshen20127** [reply]
Hi Thank you for reply. Very useful. Is there a way to output plastic strain instead of plastic deformation gradient?

**Reply by sapitts** [reply]
There is not a direct way to output the plastic strain. When I have wanted to determine the plastic strain, I output the total strain (e.g. `total_lagrangian_strain`), monitor the plastic deformation gradient diagonal terms for when the value changes from 1.0 (larger or smaller depends on the loading conditions and the location in the diagonal). I take the value of the total strain, at the timestep before the plastic deformation gradient changes from 1.0, as the elastic strain. Finally, I subtract that value of elastic strain from the total strain, at timesteps after the plastic deformation changes from 1.0, to determine the plastic strain.

Note that this method assumes no eigenstrain in the simulation.

**Reply by pshen20127** [reply]
Nice. My output elastic strain is always zero. Do you know what's wrong with my setting(please see attached input file)? And is there a way to output true stress and strain?
[polycrystal.txt](https://github.com/idaholab/moose/files/9113953/polycrystal.txt)

**Reply by sapitts** [reply]
part a) I can take a look at your input file next week but I won't have time until then

part b) Yes, use the material properties `stress` (the Cauchy stress) and `total_strain`

**Reply by pshen20127** [reply]
No problem. Thank you.

**Reply by sapitts** [reply]
Hi @pshen20127,

The elastic strain in your simulation is reported as zero because we do not save the calculated Green-Lagrange strain tensor  (defined with respect to the reference configuration) as a material property--turns out this is the same issue as with the `equivalent_slip_increment` tensor. In the crystal plasticity classes we have made the choice to not save this value to the `elastic_strain` variable because that strain measure is the Almansi strain tensor (defined with respect to the current configuration and the work conjugate of the Cauchy stress).

The elastic Green-Lagrange strain is defined locally in the `ComputeMultipleCrystalPlasticityStress` class here and the line 523 below: https://github.com/idaholab/moose/blob/4cb06f13e57bc2cccdf5f7b9439f4feb4521df66/modules/tensor_mechanics/src/materials/crystal_plasticity/ComputeMultipleCrystalPlasticityStress.C#L522
If you wanted to output that variable, you could define a new material property for that Green-Lagrange elastic strain within the `ComputeMultipleCrystalPlasticityStress`, and, if you wanted, contribute that change back to the MOOSE repository. As many members of the MOOSE team say, we accept merge requests :-)

A final note: I saw in your input file that you are using only two displacement variables, `disp_x` and `disp_y`, and thus are essentially running a 2D simulation. Are you intended to use a crystal plasticity model in a 2D simulation?

**Reply by pshen20127** [reply]
Hi sapitts, yes, it's a 2D simulation. And thanks for your deteailed and very useful explanation.
